source/ceylon/buffer/Buffer.ceylon

```diff
@@ -0,0 +1,180 @@
+     if it would otherwise exceed it (when shrinking the
+     buffer below the old [[limit]]), effectively making the
+     new [[limit]] be at the end of the buffer.
+     
+     If you are growing the array, the [[limit]] will not be
+     changed unless the optional parameter [[growLimit]] is
+     set to `true` (defaults to `false`), because when you
+     are writing it makes sense to grow the limit, but when
+     you are reading it usually does not, as growing a
+     buffer does not generate new meaningful data."
+    shared formal void resize(Integer newSize,
+        Boolean growLimit = false);
+    
+    "Returns `true` if the current [[position]] is smaller
+     than the [[limit]]."
+    shared Boolean hasAvailable => available > 0;
+    
+    "Returns the number of objects that can be read/written
+     from the current
+     [[position]] until we reach the [[limit]]."
+    shared Integer available => limit - position;
+    
+    "Reads an object from this buffer at the current
+     [[position]]. Increases the [[position]] by `1`."
+    throws (`class BufferUnderflowException`,
+        "If [[position]] >= [[limit]]")
+    shared formal Element get();
+    
+    "Writes an object to this buffer at the current
+     [[position]]. Increases the [[position]] by `1`."
+    throws (`class BufferOverflowException`,
+        "If [[position]] >= [[limit]]")
+    shared formal void put(Element element);
+    
+    "Resets the [[position]] to `0` and the [[limit]] to the
+     [[capacity]]. Use this after reading to start writing
+     to a clear buffer."
+    shared formal void clear();
+    
+    "Flips this buffer after a write operation, so that it
+     is ready to be read.
+     
+     This will and set its [[limit]] to the current
+     [[position]], and reset its [[position]] to `0`. This
+     essentially means that you will be able to read from
+     the beginning of the buffer until the last object you
+     [[put]] in it when writing."
+    shared formal void flip();
+    
+    "Returns an [[Iterator]] object to read from the current
+     [[position]] until the [[limit]]. This iterator modifies
+     the buffer directly, so you will need to [[clear]] or
+     [[flip]] it afterwards if you wish to iterate it again."
+    shared actual Iterator<Element> iterator() {
+        object it satisfies Iterator<Element> {
+            shared actual Element|Finished next() {
+                if (hasAvailable) {
+                    return get();
+                }
+                return finished;
+            }
+        }
+        return it;
+    }
+    
+    "The current underlying Array of the buffer. Any changes
+     made to the buffer will be reflected in this array, and
+     vice versa.
+     
+     Reallocations caused by a call to [[resize]] will cause any
+     prior references obtained though this attribute to become
+     stale."
+    shared formal Array<Element> array;
```

Comment:
It's important for symmetry in the `Codec` system that there's an equivalent to `String` to return from Byte-form encode/decode operations. This is the only way of producing that without copying.

I don't see `array` preventing a mapped file implementation, all that such a implementation would need to do is allocate an `Array` on-demand with a copy of the actual backing data. This isn't efficient obviously, but it wouldn't make sense to use such a buffer in the same way as `Array`-based `Buffers` are.